handler/rfc8693/handler.go

```diff
@@ -0,0 +1,269 @@
+		return errorsx.WithStack(fosite.ErrUnknownRequest)
+	}
+
+	if !requester.GetClient().GetGrantTypes().Has(string(fosite.GrantTypeTokenExchange)) {
+		return errorsx.WithStack(fosite.ErrUnauthorizedClient.WithHintf("The OAuth 2.0 Client is not allowed to use authorization grant '%s'.", fosite.GrantTypeTokenExchange))
+	}
+
+	atLifespan := fosite.GetEffectiveLifespan(requester.GetClient(), fosite.GrantTypeTokenExchange, fosite.AccessToken, c.HandleHelper.Config.GetAccessTokenLifespan(ctx))
+
+	if err := c.IssueAccessToken(ctx, atLifespan, requester, responder); err != nil {
+		return err
+	}
+
+	if canIssueRefreshToken(ctx, c, requester) {
+		refresh, refreshSignature, err := c.RefreshTokenStrategy.GenerateRefreshToken(ctx, requester)
+		if err != nil {
+			return errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+		}
+		if err := c.RefreshTokenStorage.CreateRefreshTokenSession(ctx, refreshSignature, requester); err != nil {
+			return errorsx.WithStack(fosite.ErrServerError.WithDebug(err.Error()))
+		}
+
+		responder.SetExtra("refresh_token", refresh)
+	}
+	return nil
+}
+
+func canIssueRefreshToken(ctx context.Context, c *Handler, requester fosite.Requester) bool {
+	scope := c.Config.GetRefreshTokenScopes(ctx)
+	// Require one of the refresh token scopes, if set.
+	if len(scope) > 0 && !requester.GetGrantedScopes().HasOneOf(scope...) {
+		return false
+	}
+	// Do not issue a refresh token to clients that cannot use the refresh token grant type.
+	if !requester.GetClient().GetGrantTypes().Has("refresh_token") {
+		return false
+	}
+	return true
+}
+
+func (c *Handler) CanSkipClientAuth(ctx context.Context, requester fosite.AccessRequester) bool {
+	if s := c.Config.GetGrantTypeTokenExchangeCanSkipClientAuth(ctx); s != nil {
+		return s(ctx, requester)
+	}
+
+	return false
+}
+
+func (c *Handler) keyFunc(ctx context.Context) jwt.Keyfunc {
+	return jwt.Keyfunc(func(t *jwt.Token) (interface{}, error) {
+		kid, ok := t.Header["kid"].(string)
+		if !ok {
+			return nil, errors.New("invalid kid")
+		}
+		iss, ok := t.Claims["iss"].(string)
+		if !ok {
+			return nil, errors.New("invalid iss")
+		}
+		return c.Storage.GetIDTokenPublicKey(ctx, iss, kid)
+	})
+}
+
+func (c *Handler) verifyAccessTokenAsSubjectToken(ctx context.Context, clientID string, params *tokenExchangeParams) (fosite.Requester, error) {
+	sig := c.HandleHelper.AccessTokenStrategy.AccessTokenSignature(ctx, params.subjectToken)
+	or, err := c.HandleHelper.AccessTokenStorage.GetAccessTokenSession(ctx, sig, nil)
+	if err != nil {
+		return nil, errorsx.WithStack(fosite.ErrRequestUnauthorized.WithWrap(err).WithDebug(err.Error()))
+	} else if err := c.AccessTokenStrategy.ValidateAccessToken(ctx, or, params.subjectToken); err != nil {
+		return nil, err
+	}
+
+	allowClientIDs, err := c.Storage.GetAllowedClientIDs(ctx, clientID)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, cid := range allowClientIDs {
+		if or.GetClient().GetID() == cid {
```

Comment:
Shouldn't this condition be `clientID == cid`? If my understanding is right, you want to check if the requesting client is allowed to exchange the subject token issued to another client. `or.GetClient()` will return the subject token's client?

Reply:
One other thing you might consider adding is a check to prevent a client from using token exchange to replace refresh token flow. The purpose of token exchange is for a different client to exchange one token for another. I am not sure this is in the spec but I feel it should be added. 

```
        tokenClientID := or.GetClient().GetID()
	// forbid original subjects client to exchange its own token
	if clientID == tokenClientID {
		return nil, errors.WithStack(fosite.ErrRequestForbidden.WithHint("Clients are not allowed to perform a token exchange on their own tokens."))
	}
```

The alternative is to add a Storage function to perform any additional validation. Unfortunately, adding this in ValidateAccessToken won't be practical because it is lacking the necessary parameters.

Reply:
> Shouldn't this condition be clientID == cid?

Yes, in followup with the other comment, it should be `clientID == cid`.

> One other thing you might consider adding is a check to prevent a client from using token exchange to replace refresh token flow. The purpose of token exchange is for a different client to exchange one token for another. I am not sure this is in the spec but I feel it should be added.

Yes this point is quite important, I gave it a thought but since it is not mentioned in the spec, I left it to the respective implementations. Overall I don't expect the storage to return `allowedClientIDs` containing the client for the subject token itself. But since it is not in the spec, I don't intend to impose it.